### DISCUSSION TITLE
Change ejabberd hostname in existing database and config

### DISCUSSION THREAD

**Discussion by kaizushi** [processone/ejabberd#3670] [Questions] [answered]
I want to migrate an ejabberd server from its old hostname to the new hostname. To do this I intend to just use search and replace on the configuration file for my server and also the database.

It is the  mnesia database where I do not know how to do this. I just want to know how to do a search and replace on the database so everything that matches olddatabasehostname.domain becomes newdatabasehostname.domain, and obviously there I am using a figurative hostname as I do not wish to reveal the actual hostname.

Obviously for the configuration file it is as simple as `sed -i 's/olddatabasehostname.domain/newdatabasehostname.domain/g /etc/jabber/ejabberd.yml` and I simply wish to do much the same to the database files. I really should not be doing that with sed and merely wish for some way to do it within the mnesia database.

I have tried a variety of Google searches and could not find anything complete. This problem also means my server does not work properly, long story short I have lost it's domain.

**Comment by badlop** [comment, marked as answer]
So, you changed one XMPP host in the 'hosts' option of 
https://docs.ejabberd.im/admin/configuration/basic/#host-names

How to change the JID of users (in table passwd) and the contacts (in table roster), and many other mnesia tables:

If your database is small (a few hundreds or thousands of users), you can:
1. Dump the Mnesia database to a text file using ejabberdctl or WebAdmin with
https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#dump
2. Modify that text file (with sed or whatever), making sure to keep the erlang syntax in that file
3. Clean the mnesia database (better to backup it first, of course)
4. Load your modified text file using ejabberdctl or WebAdmin with https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#load

If your database is big,  that dump and load may be extremely slow and consume all your RAM, and we would need to think another method...

Please note one obvious thing: if your users have contacts in other remote servers, they obviously won't know about this change. It is your users duty to contact them and tell that the address have changed.

**Reply by kaizushi** [reply]
I am not really sure how to use your advice. I do not have the WebAdmin and ejabberctl is not doing what I expect.

I ran the command ejabberdctl dump REDACTED-database.txt and it took some time like it was doing something. Yet the file I expected to exist of the dump `REDACTED-database.txt' does not exist.

I also tried `ejabberdctl dump /root/REDACTED-dump.txt` and got this error...
```
Can't store dump in "/root/REDACTED-dump.txt" at node ejabberd@localhost: eacces
```

I suspect that it does not like to do things in /root and therefore the dump I made the first time might just be somewhere else on my system.

**Reply by kaizushi** [reply]
I managed to get this command to work: `ejabberdctl dump /tmp/infantile-dump.txt`

I have done the search and replace required, yet I cannot find on Google how to clean the mnesia database.

**Reply by badlop** [reply]
> how to clean the mnesia database

I can think two ways:

A) Filesystem way: stop ejabberd, delete the [Mnesia database spool files](https://docs.ejabberd.im/admin/installation/#installation) (or better, move away but keeping them, just in case...). Then start ejabberd (an empty mnesia database will be created), and you can load your txt file.

B) The ejabberdctl way, using [delete_mnesia command](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#delete-mnesia)

**Reply by kaizushi** [reply]
Thank you, I actually got it working but went to bed. I tried something because I have backups anyway of the whole system that does this. I did the dump, the regexp change, and a load. The new domain is working and the config file no longer mentions the old. The old domain has stuff in my database, but I use RAM copy for everything but login credentials. There are some skeletons of the old domain in the database, but because the config file no longer refers to them all is working well.

Thanks so much, and I would be happy to tip you a little XMR.
